src/BotPlutusInterface/Balance.hs

```diff
@@ -65,71 +73,111 @@ import Plutus.V1.Ledger.Api (
       privKeys <- newEitherT $ Files.readPrivateKeys @w pabConf
-      let utxoIndex = fmap Tx.toTxOut utxos <> unBalancedTxUtxoIndex unbalancedTx
+
+      let utxoIndex :: Map TxOutRef TxOut
+          utxoIndex = fmap Tx.toTxOut utxos <> unBalancedTxUtxoIndex unbalancedTx
+
+          requiredSigs :: [PubKeyHash]
           requiredSigs = map Ledger.unPaymentPubKeyHash $ Map.keys (unBalancedTxRequiredSignatories unbalancedTx)
 
+      lift $ printBpiLog @w Debug $ viaShow utxoIndex
+
+      -- We need this folder on the CLI machine, which may not be the local machine
+      lift $ createDirectoryIfMissingCLI @w False (Text.unpack "pcTxFileDir")
+
       tx <-
         newEitherT $
           addValidRange @w
             (unBalancedTxValidityTimeRange unbalancedTx)
             (unBalancedTxTx unbalancedTx)
 
-      lift $ printBpiLog @w Debug $ viaShow utxoIndex
-
-      -- We need this folder on the CLI machine, which may not be the local machine
-      lift $ createDirectoryIfMissingCLI @w False (Text.unpack pabConf.pcTxFileDir)
-
-      -- Adds required collaterals, only needs to happen once
-      -- Also adds signatures for fee calculation
-      preBalancedTx <- hoistEither $ addTxCollaterals utxoIndex tx >>= addSignatories ownPkh privKeys requiredSigs
+      -- Adds required collaterals in the `Tx`, if `bcHasScripts`
+      -- is true. Also adds signatures for fee calculation
+      preBalancedTx <-
+        if bcHasScripts balanceCfg
+          then
+            maybe
+              (throwE "Tx uses script but no collateral was provided.")
+              (hoistEither . addSignatories ownPkh privKeys requiredSigs . flip addTxCollaterals tx)
+              mcollateral
+          else hoistEither $ addSignatories ownPkh privKeys requiredSigs tx
 
       -- Balance the tx
-      (balancedTx, minUtxos) <- loop utxoIndex privKeys [] preBalancedTx
+      (balancedTx, minUtxos) <- balanceTxLoop utxoIndex privKeys [] preBalancedTx
 
       -- Get current Ada change
       let adaChange = getAdaChange utxoIndex balancedTx
-      -- If we have change but no change UTxO, we need to add an output for it
-      -- We'll add a minimal output, run the loop again so it gets minUTxO, then update change
+          bTx = fst <$> balanceTxLoop utxoIndex privKeys minUtxos (addOutput changeAddr balancedTx)
+
+      -- Checks if there's ada change left, if there is then we check
+      -- if `bcSeparateChange` is true, if this is the case then we create a new UTxO at
+      -- the changeAddr.
       balancedTxWithChange <-
-        if adaChange /= 0 && not (hasChangeUTxO changeAddr balancedTx)
-          then fst <$> loop utxoIndex privKeys minUtxos (addOutput changeAddr balancedTx)
-          else pure balancedTx
+        case adaChange /= 0 of
+          True | bcSeparateChange balanceCfg -> bTx
+          True | not (hasChangeUTxO changeAddr balancedTx) -> bTx
+          _ -> pure balancedTx
```

Comment:
what about the following warnings that arises when using `otherwise` instead of a wildcard
1. This binding for ‘otherwise’ shadows the existing binding defined in Prelude
2. Defined but not used: ‘otherwise’
3. Hlint:    `Found: otherwise
                 Why not: _`
should I ignore them?

Fixed, the conditional logic by adding `||`.

Reply:
yeah I think you can only use otherwise in the guards, not the root case

Reply:
Oh right, this isn't even a guard, ignore this :)